src/backend/backend.c

```diff
@@ -55,6 +60,59 @@ region_t get_damage(session_t *ps, bool all_damage) {
 	return region;
 }
 
+void get_resized_damage(session_t *ps, int stacking_rank, bool whole_screen,
+                        region_t *reg_damage, region_t *reg_paint) {
+	*reg_damage = get_damage(ps, whole_screen);
+
+	if (whole_screen) {
+		pixman_region32_init(reg_paint);
+		pixman_region32_copy(reg_paint, reg_damage);
+		return;
+	}
+
+	if (!pixman_region32_not_empty(reg_damage)) {
+		// Empty means there is nothing to paint, so we don't need to resize
+		// damage.
+		pixman_region32_init(reg_paint);
+		return;
+	}
+
+	/// The adjusted damaged regions
+	assert(ps->o.blur_method != BLUR_METHOD_INVALID);
+	if (ps->o.blur_method != BLUR_METHOD_NONE && ps->backend_data->ops->get_blur_size) {
+		int blur_width, blur_height;
+		ps->backend_data->ops->get_blur_size(ps->backend_blur_context,
+		                                     &blur_width, &blur_height);
+
+		// The region of screen a given window influences will be smeared
+		// out by blur. With more windows on top of the given window, the
+		// influences region will be smeared out more.
+		//
+		// Also, blurring requires data slightly outside the area that needs
+		// to be blurred. The more semi-transparent windows are stacked on top
+		// of each other, the larger the area will be.
+		//
+		// Instead of accurately calculate how much bigger the damage
+		// region will be because of blur, we assume the worst case here.
+		// That is, the damaged window is at the bottom of the stack, and
+		// all other windows have semi-transparent background
+		int resize_factor = 1;
+		if (stacking_rank) {
+			resize_factor = stacking_rank;
+		}
+		// TODO(yshui): maybe we don't need to resize reg_damage, only reg_paint?
+		resize_region_in_place(reg_damage, blur_width * resize_factor,
+		                       blur_height * resize_factor);
+		*reg_paint = resize_region(reg_damage, blur_width * resize_factor,
+		                           blur_height * resize_factor);
+		pixman_region32_intersect(reg_paint, reg_paint, &ps->screen_reg);
+		pixman_region32_intersect(reg_damage, reg_damage, &ps->screen_reg);
+	} else {
+		pixman_region32_init(reg_paint);
+		pixman_region32_copy(reg_paint, reg_damage);
+	}
+}
```

Comment:
am i missing something or you can replace all these branches with
```c
if (ps->o.blur_method == BLUR_METHOD_NONE || !ps->backend_data->ops->get_blur_size) {
    pixman_region32_init(reg_paint);
    if (!pixman_region32_not_empty(reg_damage)) {
        pixman_region32_copy(reg_paint, reg_damage);
    }

    return;
}
```

Reply:
what do you mean?

Reply:
```c
void get_resized_damage(session_t *ps, int stacking_rank, bool whole_screen,
                        region_t *reg_damage, region_t *reg_paint) {
	*reg_damage = get_damage(ps, whole_screen);

    if (ps->o.blur_method == BLUR_METHOD_NONE || !ps->backend_data->ops->get_blur_size) {
        pixman_region32_init(reg_paint);
        if (pixman_region32_not_empty(reg_damage)) {
            pixman_region32_copy(reg_paint, reg_damage);
        }

        return
    }

	assert(ps->o.blur_method != BLUR_METHOD_INVALID);
    int blur_width, blur_height;
    ps->backend_data->ops->get_blur_size(ps->backend_blur_context,
                                            &blur_width, &blur_height);
    int resize_factor = 1;
    if (stacking_rank) {
        resize_factor = stacking_rank;
    }
    resize_region_in_place(reg_damage, blur_width * resize_factor,
                            blur_height * resize_factor);
    *reg_paint = resize_region(reg_damage, blur_width * resize_factor,
                                blur_height * resize_factor);
    pixman_region32_intersect(reg_paint, reg_paint, &ps->screen_reg);
    pixman_region32_intersect(reg_damage, reg_damage, &ps->screen_reg);
}
```
p.s. removed comments to keep it short